### PR TITLE
Add DbgEng based debug backend for Windows ##debug

### PR DESCRIPTION
**Pull request by GustavoLCR**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

This adds support for using `dbgeng.dll` (The same backend that WinDbg uses) for debugging Windows->Windows.

It uses the same command-line options as [WinDbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/windbg-command-line-options) (but doesn't support all of the for now)
 
Also adds debug register profiles for Windows ARM/ARM64

**Test plan**

Manually test that it works

### CODE REVIEW

**Comment by GustavoLCR** [comment]
Also, it would be nice to have something like https://github.com/radareorg/radare2/pull/12064 to integrate symbol info from dbgeng as flags

**Comment by XVilka** [comment]
I wonder, if it's possible to test somehow on AppVeyor or GitHub Actions Windows CI

**Comment by XVilka** [comment]
AppVeyor [complains a lot](https://ci.appveyor.com/project/radareorg/radare2/builds/34734965/job/ey93il2dykg0foru):
```
c:\projects\radare2\libr\io\p\io_dbgeng.c(42): error C2061: syntax error: identifier 'PDEBUG_CLIENT6' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(43): error C2061: syntax error: identifier 'dbgCtrl' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(43): error C2059: syntax error: ';' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(47): error C2061: syntax error: identifier 'dbgSymbols' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(47): error C2059: syntax error: ';' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(48): error C2061: syntax error: identifier 'dbgAdvanced' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(48): error C2059: syntax error: ';' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(49): error C2059: syntax error: '}' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(120): error C2061: syntax error: identifier 'DbgEngContext' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(120): error C2059: syntax error: '}' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(121): error C2061: syntax error: identifier 'DbgEngContext' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(121): error C2059: syntax error: '}' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(122): error C2061: syntax error: identifier 'DbgEngContext' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(122): error C2059: syntax error: '}' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(144): error C2065: 'PDEBUG_EVENT_CALLBACKS_IMPL': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(144): error C2146: syntax error: missing ';' before identifier 'impl' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(144): error C2065: 'impl': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(144): error C2146: syntax error: missing ';' before identifier 'This' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(145): error C2065: 'impl': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(145): error C2223: left of '->m_idbg' must point to struct/union [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(167): error C2065: 'PDEBUG_INPUT_CALLBACKS_IMPL': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(167): error C2146: syntax error: missing ';' before identifier 'impl' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(167): error C2065: 'impl': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(167): error C2146: syntax error: missing ';' before identifier 'This' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(168): error C2065: 'DbgEngContext': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(168): error C2065: 'idbg': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(168): error C2065: 'impl': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(168): error C2223: left of '->m_idbg' must point to struct/union [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(169): error C2065: 'idbg': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(169): error C2223: left of '->dbgCtrl' must point to struct/union [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(173): error C2065: 'idbg': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(173): error C2223: left of '->dbgCtrl' must point to struct/union [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(190): error C2065: 'PDEBUG_EVENT_CALLBACKS_IMPL': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(190): error C2146: syntax error: missing ';' before identifier 'impl' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(190): error C2065: 'impl': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(190): error C2146: syntax error: missing ';' before identifier 'This' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(190): error C2223: left of '->m_ref' must point to struct/union [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(190): error C2168: '_InterlockedIncrement': too few actual parameters for intrinsic function [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(191): error C2065: 'PDEBUG_INPUT_CALLBACKS_IMPL': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(191): error C2146: syntax error: missing ';' before identifier 'impl' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(191): error C2065: 'impl': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(191): error C2146: syntax error: missing ';' before identifier 'This' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(191): error C2223: left of '->m_ref' must point to struct/union [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(191): error C2168: '_InterlockedIncrement': too few actual parameters for intrinsic function [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(192): error C2065: 'PDEBUG_OUTPUT_CALLBACKS_IMPL': undeclared identifier [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
c:\projects\radare2\libr\io\p\io_dbgeng.c(192): error C2146: syntax error: missing ';' before identifier 'impl' [C:\projects\radare2\build\libr\io\4d97eb2@@r_io@sta.vcxproj]
....
```

**Comment by trufae** [comment]
So many uri handlers. What about just handling windbg://<whatever>/args like frida does? Otherwise it makes it confusing and polutes the uri space. I wont easily know which plugin is behind each of this uris

**Comment by XVilka** [comment]
@GustavoLCR can you please add a few paragraphs on how to use it in the Radare2 book? Otherwise it's hard to understand for someone without WinDbg experience.

https://github.com/radareorg/radare2book/blob/master/src/debugger/windbg.md

**Comment by XVilka** [comment]
@karliss @ITAYC0HEN @yossizap this would be nice to have in the Cutter debugger as well. Please take a look.

**Comment by ITAYC0HEN** [comment]
Good work Gustavo!
I agree with @trufae, it is indeed another option to add to the debug plugins which makes it easier to be lost. Moreover, "DbgEng" isn't a very popular name and people will more likely search and use the Windbg one.

The addition is great, but what is the best way to present it?

My question is, when coming to practice, what are the differences between the WinDbg one and DbgEng? Maybe one is for remote debugging and the other is local? If so, can be solved by using names that indicates it

What do you think?

**Comment by GustavoLCR** [comment]
> My question is, when coming to practice, what are the differences between the WinDbg one and DbgEng? Maybe one is for remote debugging and the other is local? If so, can be solved by using names that indicates it

Main ones would be that this also supports remote user-mode debugging and minidump file targets. You can also use WinDbg commands through `=!` io_system, which is convenient.

Current `windbg` plugin only supports kernel targets so maybe it should be renamed to `kd` or `winkd`, at least on windows, and make this plugin take its place, as it is closer to what WinDbg supports.

**Comment by ITAYC0HEN** [comment]
Your suggestion sounds reasonable to me :) As long as the labels are clear, and that it is documented in the Windows debugging seciton on r2book, then I think it can take windbg's KD place.

Would love to hear more opinions from others

**Comment by XVilka** [comment]
Yes, 👍 for the renaming current remote to `winkd` and making this one `windbg`.

File: libr/debug/p/debug_winkd.c [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		if (stc->state == DbgKdExceptionStateChange) {
+			dbg->reason.type = R_DEBUG_REASON_INT;
+			reason = R_DEBUG_REASON_INT;
+			break;
+		} else if (stc->state == DbgKdLoadSymbolsStateChange) {
+			dbg->reason.type = R_DEBUG_REASON_NEW_LIB;
+			reason = R_DEBUG_REASON_NEW_LIB;
+			break;
+		}
+		R_FREE (pkt);
```
[the remark this answers is not in the document]
Comment (GustavoLCR) [reply libr/debug/p/debug_winkd.c:83]: R_FREE sets it to null after free

File: libr/io/p/io_windbg.c [hunk 2]
```diff
@@ -1,106 +1,685 @@
...
+		return S_OK;                                  \
+	} else {                                          \
+		return E_NOINTERFACE;                         \
+	}                                                 \
+}
+
+#define DECLARE_ADDREF(IFace)                         \
+static ULONG IFace##_AddRef_impl(P##IFace This) {     \
+	P##IFace##_IMPL impl = (P##IFace##_IMPL)This;     \
+	return InterlockedIncrement (&impl->m_ref);       \
```
[the remark this answers is not in the document]
Comment (GustavoLCR) [reply libr/io/p/io_windbg.c:92]: `malloc` returns an aligned address, and the struct isn't packed so its elements will be aligned inside it

[the remark this answers is not in the document]
Comment (a1ext) [reply libr/io/p/io_windbg.c:92]: according to the MSDN the pointer must be aligned to [32-bit boundary](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedincrement). I had issues with this before so I suggest you re-check that

File: libr/debug/p/debug_winkd.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (a1ext) [reply libr/debug/p/debug_winkd.c:83]: oh, found it. for some reason my IDE opened different definition, probably of `r_free`. it's ok 👍

**Review by XVilka — APPROVED** [review]
I think it's ready to be merged as is. @abcSup then you will need to rebase your KDNET PR (https://github.com/radareorg/radare2/pull/17340).

**Pull request merged**